### PR TITLE
Call reset method in next tick

### PR DESCRIPTION
**Pull request by madoar**
This PR fixes the issue described in https://github.com/madoar/angular-archwizard-demo/pull/57#issuecomment-586756052

<a href="https://gitpod.io/#https://github.com/madoar/angular-archwizard/pull/264"></a>

### CODE REVIEW

File: src/lib/components/wizard-navigation-bar.component.spec.ts [hunk 1]
```diff
@@ -64,6 +64,8 @@ describe('WizardNavigationBarComponent', () => {
   it('should show the initial step correctly', () => {
     const navBarEl = wizardTestFixture.debugElement.query(By.css('aw-wizard-navigation-bar'));
 
+    wizardTestFixture.detectChanges();
```
Comment (madoar) [inline src/lib/components/wizard-navigation-bar.component.spec.ts]: This call was needed to make the test succeed

File: src/lib/directives/wizard-step-symbol.directive.spec.ts [hunk 2]
```diff
@@ -48,17 +48,19 @@ describe('WizardStepSymbolDirective', () => {
...
 
     wizardTest = wizardTestFixture.componentInstance;
     wizard = wizardTest.wizard;
-  });
+  }));
 
   it('should create an instance', () => {
     const navigationSymbolEls = wizardTestFixture.debugElement.queryAll(By.css('aw-wizard-navigation-bar ul li .step-indicator'));
 
+    wizardTestFixture.detectChanges();
```
Comment (madoar) [inline src/lib/directives/wizard-step-symbol.directive.spec.ts]: This call was needed to make the test succeed

File: src/lib/directives/wizard-step-title.directive.spec.ts [hunk 3]
```diff
@@ -48,17 +48,19 @@ describe('WizardStepTitleDirective', () => {
...
 
     wizardTest = wizardTestFixture.componentInstance;
     wizard = wizardTest.wizard;
-  });
+  }));
 
   it('should create an instance', () => {
     const navigationLinkEls = wizardTestFixture.debugElement.queryAll(By.css('aw-wizard-navigation-bar ul li a'));
 
+    wizardTestFixture.detectChanges();
```
Comment (madoar) [inline src/lib/directives/wizard-step-title.directive.spec.ts]: This call was needed to make the test succeed

**Comment by earshinov** [comment]
That's because the initial step's `selected` flag is assigned `true` during `reset`. Without additional change detection round this change is not reflected in the DOM, hence test failures. To isolate individual test cases from this, you can add `tick` and `detectChanges` into `beforeEach`. In order to use `tick`, you will have to replace `async` with `fakeAsync` as well:

```
beforeEach(fakeAsync(() => {
    wizardTestFixture = TestBed.createComponent(WizardTestComponent);
    wizardTestFixture.detectChanges();

    wizardTest = wizardTestFixture.componentInstance;
    wizard = wizardTest.wizard;

    tick();
    wizardTestFixture.detectChanges();
  }));
```

**Comment by madoar** [comment]
@earshinov thanks again for your help.
I've now also added a test case to help us identify similar issues in the future. 
After taking another look at the code I'm thinking about whether we should maybe move the `setTimeout` call from:

https://github.com/madoar/angular-archwizard/blob/223c1ab329c48ea08ccfb4a313a661dfc72a97b6/src/lib/components/wizard.component.ts#L175

to around:

https://github.com/madoar/angular-archwizard/blob/47f91cf8f7174d69233b5ccd3db8daa988f18fbd/src/lib/navigation/base-navigation-mode.interface.ts#L176

In my opinion this would be more granular because I think in theory we only need to move the "user code", i.e. everything the user provides via inputs and outputs to the wizard or its steps, to the next tick.

What do you think @earshinov?

**Comment by earshinov** [comment]
Hi! Sorry for a late reply. I would leave the `setTimeout` as is because other actions you are doing in `NavigationMode.reset` might, I suspect, cause `ExpressionChangedAfterItHasBeenChecked` error as well, for example these assignments:

```
public reset(wizard: WizardComponent): void {
    this.ensureCanReset(wizard);

    // reset the step internal state
    wizard.wizardSteps.forEach(step => {
        step.completed = step.initiallyCompleted;       // <<<
        step.selected = false;                          // <<<
        step.editing = false;                           // <<<
    });

    // set the first step as the current step
    wizard.currentStepIndex = wizard.defaultStepIndex;  // <<<
    wizard.currentStep.selected = true;                 // <<<
    wizard.currentStep.enter(MovingDirection.Forwards); // <<<
}
```

**Comment by madoar** [comment]
> I would leave the setTimeout as is because other actions you are doing in NavigationMode.reset might, I suspect, cause ExpressionChangedAfterItHasBeenChecked error as well, for example these assignments:

Good point!

**Pull request merged**